Add org-attach feature

broken-pen opened the pull request:
Hi, apologies for dropping such a large PR in your lap!

I've been working on-and-off on [attachments](https://orgmode.org/manual/Attachments.html) and it's in a state now where I believe it works as intended and very similarly to the Emacs version. :slightly_smiling_face: 

The feature seemed simple., but ended up ballooning because:

1. the [Emacs version](https://github.com/emacs-mirror/emacs/blob/master/lisp/org/org-attach.el) uses a lot of Emacs machinery that nvim doesn't have ([recursive copying](https://www.gnu.org/software/emacs/manual/html_node/elisp/Create_002fDelete-Dirs.html#index-copy_002ddirectory), [downloads](https://github.com/emacs-mirror/emacs/blob/master/lisp/org/org-attach.el#L247));
2. I ended up having to modify several utils functions (like `orgmode.utils.fs.substitute_path`) for my use case.

I've split commits up as much as I could to make accepting/rejecting each individual change easier. I've also split the implementation of org-attach into 6-7 commits to make it easier to review them.

To be clear, I don't expect all changes to go into the main repository. Let me know if there are any design decisions that you'd rather not commit to maintaining. :wink: 

Some of my more dubitable choices:
1. I wrapped a lot of `vim.uv.fs_*` functions in a module `orgmode.attach.fs`. I simply couldn't figure out how to write an async recursive copy any other way.
2. I wrote a few new dialogs in `orgmode.attach.ui`. Some look weird because they replicate Emacs dialogs (`yes_or_no_or_cancel_slow()` is the equivalent of [yes-or-no-p](https://www.gnu.org/software/emacs/manual/html_node/elisp/Yes_002dor_002dNo-Queries.html#index-yes_002dor_002dno_002dp)) and might not necessarily make sense in nvim.
3. The Emacs version _wildly_ interleaves logic and user interaction. Elisp has a lot of macros that make this work, unlike Lua. To keep the code clear, I've put the actual logic into `orgmode.attach.core` and wrapped it in `orgmode.attach`. This separation required passing around a few callbacks and effectively writing every function twice. It may be more work than it's worth.

Thanks again for keeping this project going, it's been helping me a lot!

kristijanhusak commented:
Hey! Thanks for the PR!
I still didn't look into it completely, just took a brief look.
I noticed that as part of this PR you added properties inheritance.

To make things simpler, can we extract this into its own PR? 
The same applies to anything else that I maybe missed, and is not strictly related to attachments.

troiganto commented:
Sure thing :+1: I'll mark this PR as a draft in the meantime, to keep things organized

kristijanhusak commented:
@troiganto I merged the other PR and rebased this one to resolve conflicts.
If there's nothing else to add, move this to "ready to review" and I'll start looking at it.

kristijanhusak commented:
@troiganto can this be reviewed or there are still pending changes?

troiganto commented:
Hi @kristijanhusak ! Thanks for following up, I've been racking my brain how to split the remaining changes up so that they make sense to a reviewer.

The current set of commits is the best I could do. It would be nice if you could at least look over the very first commit. I consider it an MVP, but it very obviously is not a complete implementation (misses e.g. deleting/opening/linking attachments). Let me know how you would handle this PR going forward, eg whether to split it further up.

I'm traveling the next two weeks, so I've got a bit less time, so response times should be a bit longer, but otherwise can make any changes that are necessary or desirable.

kristijanhusak commented:
I appreciate the effort to split this into smaller chunks for reviewing, it helps tremendously. I also understand that there's probably the point where splitting is really hard to achieve.
If the whole PR is completed, I'll look at it, and leave feedback. If it's mostly good (which looks like it at the first glance), I'll probably merge it and address smaller things myself, and any bigger things we can discuss as an improvement in the next PRs. 
If that sounds good, let me know, and I'll look into this in the following days.
Thanks!

troiganto commented:
Sounds good to me!

I've also gone over the commits once more and added notes to the commit messages, which Emacs functions are ported in which commit. This makes the reasoning behind each addition (hopefully!) a bit clearer.

kristijanhusak reviewed (changes requested):
Very impressive work! 

I left a few comments, and gave it a quick test.
It seems we need to address some minor things:

1. We need to wrap something into vim.schedule 
2. Seems like non-absolute links are not working correctly

Here's a video for reference:

https://github.com/user-attachments/assets/d8125f6d-b5a5-4112-a5c2-765f84a80ad6

broken-pen committed:
feat(global): add `:Org attach` subcommand

This brings the feature in line with capture and agenda, which similarly
have a top-level menu prompt.

broken-pen committed:
feat(attach): add config `org_attach_dir_relative`

broken-pen committed:
feat(attach): add `OrgAttach:expand_links()`

This corresponds to the Emacs function `org-attach-expand-links`.

broken-pen committed:
feat(attach): add `OrgAttach:attach_url()`

This corresponds to the Emacs function `org-attach-url`. It also adds
some functionality from the Emacs core that we don't get for free.

broken-pen committed:
feat(attach): add `OrgAttach:delete_one()` and `delete_all()`

These correspond to the following Emacs functions:
- `org-attach-delete-one`
- `org-attach-delete-all`

broken-pen committed:
feat(attach): add ability to open attachments

This adds functions that correspond to these Emacs functions:

- `org-attach-reveal`
- `org-attach-reveal-in-emacs`
- `org-attach-open`
- `org-attach-open-in-emacs`

broken-pen committed:
feat(attach): add org-attach feature

The implementation follows the Emacs implementation in spirit, with some
additional separation of concerns:

1. the actual logic is implemented in `attach/core.lua`, with no
   concerns for UI or configuration;
2. `attach/init.lua` combines the core with UI and configuration and
   provides the public API;
3. `attach/ui.lua` contains any dialogs needed by the module;
4. `attach/fileops.lua` contains file operations that are provided to
   the Emacs implementation by the Emacs core (e.g. recursive
   copy/deletion of directories)
5. `attach/node.lua` provides an abstraction over headlines and whole
   files that is absent in the Emacs implementation
6. `attach/translate_id.lua` corresponds to the Emacs implementation's
   functions `org-attach-id-uuid-folder-format`,
   `org-attach-id-ts-folder-format`, and
   `org-attach-id-fallback-folder-format`. They are separated like this
   because referring to pre-defined functions in Lua is more difficult
   than in Emacs.

To reduce complexity, the following functions are left unimplemented in
this commit:

- `org-attach-file-list`
- `org-attach-expand`
- `org-attach-follow`
- `org-attach-complete-link`
- `org-attach-reveal`
- `org-attach-reveal-in-emacs`
- `org-attach-open`
- `org-attach-open-in-emacs`
- `org-attach-delete-one`
- `org-attach-delete-all`
- `org-attach-sync`
- `org-attach-archive-delete-maybe`
- `org-attach-expand-links`
- `org-attach-url`
- `org-attach-dired-to-subtree`

broken-pen committed:
feat(attach): add `OrgAttach:sync()`

This corresponds to the Emacs function `org-attach-sync`.

broken-pen committed:
feat(attach): add event to delete attachments of archived headlines

This corresponds to the Emacs function
`org-attach-archive-delete-maybe`.

broken-pen committed:
feat(attach): add attachment links

This provides functionality provided by the following functions in the
Emacs implementation:

- `org-attach-file-list`
- `org-attach-expand`
- `org-attach-follow`
- `org-attach-complete-link`

broken-pen committed:
feat(attach): add `OrgAttach:attach_to_other_buffer()`

This function has no direct Emacs equivalent. However, it tries to
replicate the functionality provided by `org-attach-dired-to-subtree`,
which attaches files selected in a dired window (roughly equivalent to
Vim's NetRW) to the last-seen Orgmode task.

Our implementation tries to be extremely general because the Neovim
ecosystem has a plethora of file browser plugins.

broken-pen committed:
feat(attach): add attachment events

These are implemented in Emacs as the hooks
`org-attach-after-change-hook` and `org-attach-open-hook`.

broken-pen commented on lua/orgmode/events/listeners/attach_maybe_delete_archived.lua:
```diff
@@ -0,0 +1,4 @@
+---@param event OrgHeadlineArchivedEvent
+return function(event)
+  require('orgmode.attach'):maybe_delete_archived(event.headline)
```
Done!

broken-pen commented on lua/orgmode/attach/node.lua:
```diff
@@ -0,0 +1,251 @@
+
+---Remove the `org_attach_auto_tag`.
+function AttachNode:remove_auto_tag()
+  -- TODO: There is currently no way to set #+FILETAGS programmatically. Do
```
Done!

broken-pen commented on lua/orgmode/attach/fileops.lua:
```diff
@@ -0,0 +1,382 @@
+---@return OrgPromise<string> tmpfile
+local function netrw_read(url)
+  return Promise.new(function(resolve, reject)
+    if not vim.g.loaded_netrwPlugin then
```
I've removed `netrw_read()` and replaced `download_file()` with a version that simply calls `curl` via `vim.system()`. That's very similar to what's proposed in https://github.com/neovim/neovim/pull/33964/ as `vim.net.download()`, so I hope it'll suffice.

broken-pen commented:
Hi! Sorry for the delay, I've been pretty busy transitioning these past few months :sweat_smile: but I finally found time to look at this PR again.

I've addressed all your suggestions and fixed the two bugs in the video. (Thanks for that, btw, it made it very easy to locate the issue!)

I think there are a few more TODOs I've left in the code. It'd be cool if you could have a look over them and suggest how to handle/document them before merging, just so they won't get lost.

kristijanhusak commented on docs/configuration.org:
```diff
@@ -2790,6 +2993,43 @@ Running [[#org_babel_tangle][org_babel_tangle]] will create file =~/org/my_tangl
+most recently open org file:
+
+#+begin_src lua
+vim.keymap.set('n', '<Leader>o+', function()
```
I wasn't able to get this working. I get this error:
```
E5108: Error executing lua: /home/kristijan/github/orgmode/lua/orgmode/attach/core.lua:143: attempt to call a table value                                                                                                                                                                                                                                                                     
stack traceback:                                                                                                                                                                                                                                                                                                                                                                              
        /home/kristijan/github/orgmode/lua/orgmode/attach/core.lua:143: in function 'list_current_nodes'                                                                                                                                                                                                                                                                                      
        /home/kristijan/github/orgmode/lua/orgmode/attach/init.lua:698: in function 'find_other_node'                                                                                                                                                                                                                                                                                         
        /home/kristijan/github/orgmode/lua/orgmode/attach/init.lua:624: in function 'attach_to_other_buffer'                                                                                                                                                                                                                                                                                  
        .../kristijan/.config/nvim/lua/partials/plugins/orgmode.lua:176: in function <.../kristijan/.config/nvim/lua/partials/plugins/orgmode.lua:173> 
```

I might not understand how it should be used. This is what I did:
1. Had a `todos.org` file with a headline and a link to a picture in the headline content. Something like this:
```
* Test
   [[~/path/to/image.png]]
```
2. Opened another `notes.org` file that has a `* Headline` inside
3. Went back to `todos.org`, hovered over the link, and did `<leader>o+`

Additionally, I would prefer if we would not suggest using the `orgmode` instance directly but instead expose this through the API if possible.

kristijanhusak reviewed (commented):
Thanks for the changes! I now managed to give it a solid test. I might not cover everything but I tried to go through the most important flows. 
Note that I'm not a user of attachments in Orgmode so I might lack some knowledge around it, but I did a comparison with how Emacs works to give a feedback.

These are the things I found:
1. Do not prompt if there's only a single file attached on the node
2. Attaching from URL throws this error: `[orgmode] .../kristijan/github/orgmode/lua/orgmode/attach/fileops.lua:362: attempt to call field 'exists' (a nil value)`.
   I tried this url https://placecats.com/millie/300/150
3. Pressing `Esc` in "Set directory" prompt shows this error `[orgmode] /home/kristijan/github/orgmode/lua/orgmode/attach/ui.lua:24: attempt to index local 'answer' (a nil value)`.
   We should just print "Quit" in that case.
4. When setting and unsetting the directory, copying over overwrites the files. In emacs, it throws a "File exists" error. Here are steps to reproduce it:
   1. Have some attachments for a headline
   2. Set the directory, when asked to copy, say "yes", when asked to delete, say "no"
   3. Unset the directory, when asked to copy, say "yes"
   4. In here, it overwrites the files
   5. In emacs, throws "File exists" error
5. Attaching a buffer content throws the `[orgmode] /home/kristijan/github/orgmode/lua/orgmode/attach/core.lua:504: E5560: Vimscript function must not be called in a fast event context`.
   It attaches the file, but also throws the error.
   Here's how I reproduce it:
   1. Have 2 org files, todos.org and notes.org
   2. Open todos.org with Neovim `nvim todos.org`
   3. Open notes.org within Neovim to load it into buffer `:e notes.org`, go back to `todos.org`
   4. On some headline, try to attach `refile.org` as a buffer content
6. Opening an attachment with `o` behaves differently. Here we are forcing an external call, where Emacs seems to open it internally most of the time.
   For example, if you attach an org file buffer and then open it, Emacs just opens another buffer and loads the org file, while we open with the external program.
   I assume it was hard to mirror that functionality in Neovim, but please confirm.
7. I wasn't able to get the completion for the attachment links to work. I'm not sure why though.


Regarding the TODOs:
1. adding events (hooks) is fine.
2. Regarding the expanding links, you should be able to capture links with `link` and `link_desc` TS node types. It was added in the recent versions.
3. Filetags can be added separately


Review is a bit all over the place, but it's a big PR so I just wrote things as I found them.
